### PR TITLE
New JointDrive Crawler environments

### PR DESCRIPTION
**Pull request by Hunter-Unity**
Replace the old crawler demo. This crawler uses the new JointDrive movement schema. Include 3 demo scenes.

### CODE REVIEW

File: unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+        if(bp.rb.transform != body)
+        {
+            AddVectorObs(Quaternion.FromToRotation(body.forward, bp.rb.transform.forward));
+        }
+    }
+
+    void Update()
+    {
+        //update pos to target every tick
+		dirToTarget = target.position - bodyParts[body].rb.position;
```
Comment (gcusso) [inline unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs]: I think this line should be in FixedUpdate to be independent of the time scale

File: unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/AdjustTrainingTimescale.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
+﻿//This script lets you change time scale during training. It is not a required script for this demo to function
+
+using System.Collections;
```
Comment (vincentpierre) [inline unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/AdjustTrainingTimescale.cs]: Should this be in the shared assets ?

Comment (vincentpierre) [reply unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/AdjustTrainingTimescale.cs]: Is it needed in the repo ?

File: unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+using UnityEngine;
+
+public class CrawlerAgent : Agent {
+    [Header("Target To Walk Towards")] 
+    [Space(10)] 
+    public Transform target;
+    public bool respawnTargetWhenTouched;
+    public float targetSpawnRadius;
+
+
```
Comment (vincentpierre) [inline unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs]: Too many blank lines

File: unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+    public MeshRenderer foot3;
+    public Material groundedMaterial;
+    public Material unGroundedMaterial;
+
+
+    /// <summary>
+    /// Used to store relevant information for acting and learning for each body part in agent.
+    /// </summary>
+    [System.Serializable]
+    public class BodyPart
```
Comment (vincentpierre) [inline unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs]: Maybe this should be its own script in the shared asset folder ?

File: unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+        if(useFootGroundedVisualization)
+        {
+            foot0.material = bodyParts[leg0_lower].groundContact.touchingGround? groundedMaterial: unGroundedMaterial;
+            foot1.material = bodyParts[leg1_lower].groundContact.touchingGround? groundedMaterial: unGroundedMaterial;
+            foot2.material = bodyParts[leg2_lower].groundContact.touchingGround? groundedMaterial: unGroundedMaterial;
+            foot3.material = bodyParts[leg3_lower].groundContact.touchingGround? groundedMaterial: unGroundedMaterial;
+        }
+    }
+
+	/// <summary>
```
Comment (vincentpierre) [inline unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs]: Indent

File: unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+    /// Adds the raycast hit dist and relative pos to observations
+    /// </summary>
+    void RaycastObservation(Vector3 pos, Vector3 dir, float maxDist)
+    {
```
Comment (vincentpierre) [inline unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs]: Is it possible to only use the Raycast component introduced in 0.3 ? Why do you need a new Raycast method ?

File: unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerContact.cs [hunk 7]
```diff
@@ -0,0 +1,64 @@
+﻿using System.Collections;
```
Comment (vincentpierre) [inline unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerContact.cs]: Is it possible to use the generic ground contact script that is used for the humanoid walker ?

File: unity-environment/Assets/ML-Agents/Examples/Crawler/crawler POST PROCESSING.asset [hunk 8]
```diff
@@ -0,0 +1,320 @@
+%YAML 1.1
```
Comment (vincentpierre) [inline unity-environment/Assets/ML-Agents/Examples/Crawler/crawler POST PROCESSING.asset]: Can you explain what this does and how much it is needed ?

File: unity-environment/Assets/ProBuilder Data/ProBuilderPreferences.asset [hunk 9]
```diff
@@ -0,0 +1,25 @@
+%YAML 1.1
```
Comment (vincentpierre) [inline unity-environment/Assets/ProBuilder Data/ProBuilderPreferences.asset]: What does `ProBuilder Data` do ?

File: unity-environment/ProjectSettings/EditorBuildSettings.asset [hunk 10]
```diff
@@ -4,4 +4,11 @@
 EditorBuildSettings:
```
Comment (vincentpierre) [inline unity-environment/ProjectSettings/EditorBuildSettings.asset]: This file should not have been pushed.

File: unity-environment/ProjectSettings/ProjectSettings.asset [hunk 11]
```diff
@@ -107,6 +107,7 @@ PlayerSettings:
   xboxOneLoggingLevel: 1
```
Comment (vincentpierre) [inline unity-environment/ProjectSettings/ProjectSettings.asset]: Does this file really need to be modified ?

File: unity-environment/ProjectSettings/ProjectVersion.txt [hunk 12]
```diff
@@ -1 +1 @@
-m_EditorVersion: 2018.1.0b6
```
Comment (vincentpierre) [inline unity-environment/ProjectSettings/ProjectVersion.txt]: This file should not have been pushed

File: unity-environment/ProjectSettings/TagManager.asset [hunk 13]
```diff
@@ -22,6 +22,7 @@ TagManager:
   - orangeBlock
   - block
   - orangeGoal
+  - crawlerTarget
```
Comment (vincentpierre) [inline unity-environment/ProjectSettings/TagManager.asset]: I believe it is better to keep these tags as generic as possible as they are shared across all scenes.

**Comment by vincentpierre** [comment]
* When checking the branch out I get the following warning :
```
Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs(36,17): 
warning CS0414: 
The private field `CrawlerAgent.academy' is assigned but its value is never used
```

 * Please do not use spaces in the names of folders or files use 
```
find . -name "* *" 
```
to get a list of the files containing spaces in their path

 * Put the scenes into a `Scenes` folder since there are more than one.

 * There is a `Crawler` in the `Crawler` prefab, not sure why.

 * There is a `TARGET` in the static target scene but I do not know what it is for. Since it is possible to move this target, users might be confused.

 * The default brain type should be Player in all the scenes.

 * I think we can get rid of the old Crawler prefab, we can always look at previous versions if we want to revert.

 * Why is there a pyramid prefab, I could not find a pyramid in the scenes.

File: unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs (hunk 6 above)
Comment (Hunter-Unity) [reply unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs]: this one is much less complex and also gets the hit position relative to the body

File: unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerContact.cs (hunk 7 above)
Comment (Hunter-Unity) [reply unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerContact.cs]: the crawler needs to be able to detect when the target is touched. we'll probably need to detect even more types of objects going forward which is why i made a new script

File: unity-environment/Assets/ML-Agents/Examples/Crawler/crawler POST PROCESSING.asset (hunk 8 above)
Comment (Hunter-Unity) [reply unity-environment/Assets/ML-Agents/Examples/Crawler/crawler POST PROCESSING.asset]: post processing profile image profile (to adjust the colors). we can get rid

File: unity-environment/Assets/ProBuilder Data/ProBuilderPreferences.asset (hunk 9 above)
Comment (Hunter-Unity) [reply unity-environment/Assets/ProBuilder Data/ProBuilderPreferences.asset]: leftover data can get rid

Comment (Hunter-Unity) [reply unity-environment/Assets/ProBuilder Data/ProBuilderPreferences.asset]: removed

File: unity-environment/Assets/ML-Agents/Examples/Crawler/crawler POST PROCESSING.asset (hunk 8 above)
Comment (Hunter-Unity) [reply unity-environment/Assets/ML-Agents/Examples/Crawler/crawler POST PROCESSING.asset]: removed

File: unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs (hunk 5 above)
Comment (Hunter-Unity) [reply unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs]: fixed

File: unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs (hunk 3 above)
Comment (Hunter-Unity) [reply unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs]: fixed

File: unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/AdjustTrainingTimescale.cs (hunk 2 above)
Comment (Hunter-Unity) [reply unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/AdjustTrainingTimescale.cs]: moved to shared assets

File: unity-environment/ProjectSettings/TagManager.asset (hunk 13 above)
Comment (Hunter-Unity) [reply unity-environment/ProjectSettings/TagManager.asset]: changed to "target"

File: unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs (hunk 3 above)
Comment (vincentpierre) [reply unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs]: I think the formatting is inconsistent. Could you reformat all the .cs scripts ? (for example new line for bracket)

File: unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs [hunk 14]
```diff
@@ -126,9 +121,11 @@ public void SetupBodyPart(Transform t)
         bp.groundContact = t.GetComponent<CrawlerContact>();
 		bp.agent = this;
     }
+
+    //Initialize
```
Comment (vincentpierre) [inline unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs]: method comments should be `///<summary>...`

File: unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs [hunk 15]
```diff
@@ -138,6 +135,9 @@ public override void InitializeAgent()
         SetupBodyPart(leg2_lower);
         SetupBodyPart(leg3_upper);
         SetupBodyPart(leg3_lower);
+
+        //we want a lower center of mass or the crawler will roll over easily. 
+        //these settings shift the COM on the lower legs
 		bodyParts[leg0_lower].rb.centerOfMass = footCenterOfMassShift;
```
Comment (vincentpierre) [inline unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs]: Indent looks weird here

File: unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs [hunk 16]
```diff
@@ -164,21 +164,21 @@ public void CollectObservationBodyPart(BodyPart bp)
...
-    }
+    // void FixedUpdate()
+    // {
+    //     //update pos to target
+	// 	dirToTarget = target.position - bodyParts[body].rb.position;
+
+    //     //if enabled the feet will light up green when the foot is grounded.
+    //     //this is just a visualization and isn't necessary for function
+    //     if(useFootGroundedVisualization)
+    //     {
```
Comment (vincentpierre) [inline unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs]: Dead code

File: unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs [hunk 17]
```diff
@@ -276,21 +287,24 @@ public override void AgentAction(float[] vectorAction, string textAction)
     void RewardFunctionMovingTowards()
     {
         //don't normalize vel. the faster it goes the more reward it should get
```
Comment (vincentpierre) [inline unity-environment/Assets/ML-Agents/Examples/Crawler/Scripts/CrawlerAgent.cs]: These comments are a little awkward...

**Comment by vincentpierre** [comment]
@Hunter-Unity regarding the commit message on commit [04315ab](https://github.com/Unity-Technologies/ml-agents/pull/777/commits/04315ab487d549c916768bfdf59fd03f281d75c1), The latest supported version of TensorFlow is 1.7.0

**Comment by xiaomaogy** [comment]
@Hunter-Unity To be specific, the latest supported version of TensorFlow is 1.7.1, see the PR in https://github.com/Unity-Technologies/ml-agents/pull/746, use the download link https://s3.amazonaws.com/unity-ml-agents/0.4/TFSharpPlugin.unitypackage and change the requirement.txt

-tensorflow==1.4.0
+tensorflow==1.7.1

To try.

**Comment by awjuliani** [comment]
This PR has been superseded by: https://github.com/Unity-Technologies/ml-agents/pull/841

**Pull request closed without merging**